app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionPresenter.kt

```diff
@@ -0,0 +1,127 @@
+        val connection = connectionsRepository.getByGuid(connectionGuid) ?: return
+        this.actionDeepLinkData = actionDeepLinkData
+        this.connectionAndKey = createConnectionAndKey(
+            connection = connection,
+            keyStoreManager = keyStoreManager
+        )
+        if (connectionAndKey == null) viewMode = ViewMode.ACTION_ERROR
+    }
+
+    override fun getTitleResId(): Int = R.string.connections_new_connection
+
+    override fun onViewCreated() {
+        if (viewMode == ViewMode.START) {
+            apiManager.sendAction(
+                actionUUID = actionDeepLinkData?.actionUuid ?: "",
+                connectionAndKey = connectionAndKey ?: return,
+                resultCallback = this
+            )
+            viewMode = ViewMode.PROCESSING
+        }
+        setupViews()
+    }
+
+    override fun onActionInitFailure(error: ApiErrorData) {
+        viewMode = ViewMode.ACTION_ERROR
+        setupViews()
+        viewContract?.showErrorAndFinish(error.getErrorMessage(appContext))
+    }
+
+    override fun onActionInitSuccess(response: ActionData) {
+        val authorizationID = response.authorizationId ?: ""
+        val connectionID = response.connectionId ?: ""
+        if (response.success == true && authorizationID.isEmpty() && connectionID.isEmpty()) {
+            viewMode = ViewMode.ACTION_SUCCESS
+            setupViews()
+        } else {
+            viewContract?.returnActionWithConnectionId(
+                authorizationIdentifier = AuthorizationIdentifier(
+                    authorizationID = authorizationID,
+                    connectionID = connectionID
+                )
+            )
+        }
+    }
+
+    override fun onViewClick(viewId: Int) {
+        if (viewId == R.id.mainActionView) {
+            viewContract?.closeView()
+        }
+    }
+
+    private fun setupViews() {
+        when (viewMode) {
+            ViewMode.ACTION_SUCCESS -> {
+                viewContract?.showCompleteView(true)
+                viewContract?.updateCompleteViewContent(
+                    iconResId = R.drawable.ic_complete_ok_70,
+                    completeTitle = appContext.getString(R.string.action_feature_title),
+                    completeMessage = appContext.getString(R.string.action_feature_description),
+                    mainActionTextResId = R.string.actions_proceed
+                )
+            }
+            ViewMode.ACTION_ERROR -> {
+                viewContract?.showCompleteView(true)
+                viewContract?.updateCompleteViewContent(
+                    iconResId = R.drawable.ic_auth_error_70,
+                    completeTitle = appContext.getString(R.string.action_error_title),
+                    completeMessage = appContext.getString(R.string.action_error_description),
+                    mainActionTextResId = R.string.actions_try_again
+                )
+            }
+            ViewMode.PROCESSING -> viewContract?.showCompleteView(false)
```

Comment:
Show progress progress explicitly.
The presenter should not know about the structure of the view.